design/007-external-data-handling/004-external-data-handling.md

```diff
@@ -0,0 +1,421 @@
+Store credentials? (y/N): y
+-- output --
+Credentials stored in ~/.renku/renku.ini
+```
+
+On the filesystem, the user would see `data/mydataset` as is normal for Renku
+datasets. Importantly, `data/mydataset` is also added to `.gitignore` to avoid
+exploding the git repository. 💥
+
+`s3://` is used here but we should also support at least `nfs://`. Renku should
+ask for credentials in a way that is appropriate for the type of resource that
+is being added.
+
+#### Getting information about the dataset:
+
+```
+renku dataset show very-large-data
+-- output --
+Name: very-large-data
+Created: 2022-03-30 15:55:59+02:00
+Creator(s): Rok Roškar <rok.roskar@sdsc.ethz.ch>
+Title: very-large-data
+Description: Data that is very large. Indeed.
+Data Sources:
+    - type: s3
+      url: s3://server/path/to/bucket
+      location: None
+```
+
+#### Staging external data:
+
+External data has to somehow be brought into the system where Renku is running.
+This seems to be a "staging" operation, but we could also call it something
+else. "Mounting" is another option but seems too linux-specific.
+
+```
+renku dataset stage very-large-data
+```
+
+This command uses Fuse when necessary to mount data locally. The command should
+default to mounting the remote filesystem, but should give the option to copy, e.g.
+
+```
+renku dataset stage --copy very-large-data
+```
+
+The default location is to be `.renku/mounts/<name>` but can be specified with
+the `--location` flag in case it is preferred to take the data to an external
+drive or another filesystem.
+
+Once it's staged, it can be found locally and used in workflows:
+
+```
+renku dataset show very-large-data
+
+Name: very-large-data
+Created: 2022-03-30 15:55:59+02:00
+Creator(s): Rok Roškar <rok.roskar@sdsc.ethz.ch>
+Title: very-large-data
+Description: Data that is very large. Indeed.
+External Sources:
+    - type: s3
+      url: s3://server/path/to/<bucket>
+      location: .renku/mounts/<bucket>
+```
+
+The data is accessible under `data/very-large-data/<bucket>` in the project's directory.
+
+If the data is already mounted by some other means, `dataset stage` can be used
+to let Renku know about it:
+
+```
+renku dataset stage --existing s3://server/path/to/<bucket>:/mounts/bucket
+```
+
+#### Using external data _without_ Renku datasets
+
+We should also consider the case where creating a dataset is too much effort or
+does not fit the stage of the project. In such cases, Renku should allow the
+user to bring in and reference external data. For this, we should review the
+existing `storage` subcommand. For example, to bring in data from an S3 bucket,
+the user would do
+
+```
+renku storage add s3://server/path/to/<bucket>
+-- output --
+<bucket> added as external storage
+```
+
+This would not immediately do anything with the data, but only create the
+`DataStorage` metadata object. For bringing in the data, the user would need
+to do
+
+```
+renku storage stage --copy <bucket> path/to/bucket
```

Comment:
The data is copied to `.renku/mounts/<bucket>` - though arguably it should be called something other than `mounts`. Without `--copy` it's mounted there directly. In both cases it's symlinked to the location in the project. But maybe that's overcomplicating it and it could just be copied/mounted directly in `path/to/bucket`?

Reply:
I mean, the whole `Using external data _without_ Renku datasets` section uses `--copy` in all the examples, doesn't mention copying in the description around those commands, and does not address that a user could also access the mount directly without a copy.

Reply:
I've reworked the copy/no-copy parts